### PR TITLE
C4 Leverage Macro Fixes

### PR DESCRIPTION
**Pull request by wtj2021**
- [x] https://github.com/code-423n4/2023-10-badger-findings/issues/323
- [x] https://github.com/code-423n4/2023-10-badger-findings/issues/151
- [x] https://github.com/code-423n4/2023-10-badger-findings/issues/112

### CODE REVIEW

**Comment by dapp-whisperer** [comment]
# Review Pt. 1
update **ISortedCdps** interface to expose `toCdpId` and `nextCdpNonce` so that consuming contracts can deterministically find cpdId of Cdp that is being created.

## LeverageMacroBase

Is now officially abstract. It was always intended to be inherited and is used this way in practice.

### New Operation: ClaimSurplusOperation

We now allow users to claim surplus as an action in the flashloan fallback of a leverage op. 

This fixes issues where positions opened via a LeverageMacro impl could have no way to get out collateral surplus out (unless the implementing contract also included that functionality separately)

### How it’s done

we have a new distinct `OperationType`, `ClaimSurplusOperation`

```solidity
enum OperationType {
        OpenCdpOperation,
        AdjustCdpOperation,
        CloseCdpOperation,
        **ClaimSurplusOperation**
    }
```

it corresponds with a new `PostOperationCheck` type, `claimSurplus`

This allows for confirming the collateral shares that were gained as part of the operat

```solidity
enum PostOperationCheck {
        openCdp,
        cdpStats,
        isClosed,
        **claimSurplus**
    }
```

we piggyback on the `PostCheckParams.expectedCollateral` parameter to verify expected output from claim operation

During a call 

The new flashloan callback is correspondingly simple, no parameters are required. We claim the outstanding surplus that belongs to the contract that has the callback context.

```jsx
function _claimSurplusCallback() internal {
        borrowerOperations.claimSurplusCollShares();
    }
```

For collSurplus post-operation checks, we read the stETH shares in the macro before and after the operation and compare them to ensure our shares increased by at least the expected amount.

we then assume all shares that entered the macro as part of the operation are from this. 

The option to sweep all collateral to the caller is retained based on the immutable `willSweep` flag

**Comment by GalloDaSballo** [comment]
Wdyt about removing the safety checks for targets
That way owner can just call w/e target and w/e signature they need

@wtj2021 @dapp-whisperer

**Comment by dapp-whisperer** [comment]
# NotSystem
In response to https://github.com/code-423n4/2023-10-badger-findings/issues/112, @GalloDaSballo mentions:

> I think we wil remove the safety check and add a way to reset allowances

does this imply you support removal of the [_ensureNotSystem check](https://github.com/ebtc-protocol/ebtc/pull/727/files#diff-e88028669a4c979ebd5796039dc6849aceeb57c343e51bdaf363caf9e4324ab1R497) to prevent arbitrary calls to the eBTC contracts?

I was wondering if you could expand further on allowance reset comment?

# My Take
The reviewer syas the following
> This will result in the inability to access the borrowerOperations.permitPositionManagerApproval method for authorization

> In the _ensureNotSystem method, remove restrictions on borrowerOperations access.

> In fact, I think that the restrictions in _ensureNotSystem can be removed. This is a wallet. Any developer can develop a similar wallet according to his own wishes. The restrictions here have little significance for security.

In creating the zaps, we do in fact get around this by wrapping the LeverageMacroBase in something that can call the positionManager functions.

So this is only really relevant for the LeverageMacroReference impl.

**Comment by dapp-whisperer** [comment]
> Wdyt about removing the safety checks for targets That way owner can just call w/e target and w/e signature they need
> 
> @wtj2021 @dapp-whisperer

I'm fine removing it as per previous comment.

**Comment by wtj2021** [comment]
@CodingNameKiki Yes, that's correct. **toCdpId** is used to replace **cdpOfOwnerByIndex**, which can cause issues for large sortedCdp lists.

**Comment by wtj2021** [comment]
- removed claimSurplus post validation
- added expectedCdpId to _doOperation (internal) for ZapRouter

**Comment by GalloDaSballo** [comment]
Overall am thinking we could add a `performArbitraryCall` operation to allow bypassing those checks and allow adding functionality we currently don't have

I would put this as low prio

I'd be interested in a cut version of the macro, that uses the same audited code, but removes some extra pieces as to make it easier to read and to understand

Overall we should look to simplify the code more so than adding a bunch of additional if/elses

**Comment by wtj2021** [comment]
@CodingNameKiki that makes sense, I added a `None` operation type for doing trades and a check to calculate the expectedCdpId if post check type is set to none.

**Review by dapp-whisperer — APPROVED** [review]

**Comment by dapp-whisperer** [comment]
@GalloDaSballo @rayeaster 

after discussion is ready to merge down to main. If you of you guys can stamp we can merge.

File: packages/contracts/contracts/Interfaces/ISortedCdps.sol [hunk 1]
```diff
@@ -82,4 +82,12 @@ interface ISortedCdps {
         bytes32 startNodeId,
         uint maxNodes
     ) external view returns (bytes32, bool);
+
+    function toCdpId(
+        address owner,
+        uint256 blockHeight,
+        uint256 nonce
+    ) external view returns (bytes32);
```
Comment (rayeaster) [inline packages/contracts/contracts/Interfaces/ISortedCdps.sol:90]: this function could be `pure` instead of `view`

**Review by rayeaster — APPROVED** [review]

**Review by dapp-whisperer — APPROVED** [review]

**Pull request merged**